util/httputil/client_test.go

```diff
@@ -0,0 +1,412 @@
+		}
+		if ExpectedPassword != password {
+			t.Errorf("The expected password (%s) differs from the obtained password (%s).", ExpectedPassword, password)
+		}
+	}, nil, nil)
+
+	//Normal flow
+	basicAuthRoundTripper := NewBasicAuthRoundTripper(ExpectedUsername,
+		ExpectedPassword, fakeRoundTripper)
+	request, _ := http.NewRequest("GET", "/hitchhiker", nil)
+	request.Header.Set("User-Agent", "Douglas Adams mind")
+	basicAuthRoundTripper.RoundTrip(request)
+
+	//Should honor already Authorization header set
+	basicAuthRoundTripperShouldNotModifyExistingAuthorization := NewBasicAuthRoundTripper(newUsername,
+		newPassword, fakeRoundTripper)
+	request, _ = http.NewRequest("GET", "/hitchhiker", nil)
+	request.SetBasicAuth(ExpectedUsername, ExpectedPassword)
+	basicAuthRoundTripperShouldNotModifyExistingAuthorization.RoundTrip(request)
+}
+
+func TestTLSConfig(t *testing.T) {
+	configTLSConfig := config.TLSConfig{
+		CAFile:             TLSCAChainPath,
+		CertFile:           BarneyCertificatePath,
+		KeyFile:            BarneyKeyNoPassPath,
+		ServerName:         "localhost",
+		InsecureSkipVerify: false}
+
+	tlsCAChain, err := ioutil.ReadFile(TLSCAChainPath)
+	if err != nil {
+		t.Fatalf("Can't read the CA certificate chain (%s)",
+			TLSCAChainPath)
+	}
+	rootCAs := x509.NewCertPool()
+	rootCAs.AppendCertsFromPEM(tlsCAChain)
+
+	barneyCertificate, err := tls.LoadX509KeyPair(BarneyCertificatePath, BarneyKeyNoPassPath)
+	if err != nil {
+		t.Fatalf("Can't load the client key pair ('%s' and '%s'). Reason: %s",
+			BarneyCertificatePath, BarneyKeyNoPassPath, err)
+	}
+
+	expectedTLSConfig := &tls.Config{
+		RootCAs:            rootCAs,
+		Certificates:       []tls.Certificate{barneyCertificate},
+		ServerName:         configTLSConfig.ServerName,
+		InsecureSkipVerify: configTLSConfig.InsecureSkipVerify}
+	expectedTLSConfig.BuildNameToCertificate()
+
+	tlsConfig, err := NewTLSConfig(configTLSConfig)
+	if err != nil {
+		t.Fatalf("Can't create a new TLS Config from a configuration (%s).", err)
+	}
+
+	if !reflect.DeepEqual(tlsConfig, expectedTLSConfig) {
+		t.Fatalf("Unexpected TLS Config result: \n\n%+v\n expected\n\n%+v", tlsConfig, expectedTLSConfig)
+	}
+}
+
+func TestTLSConfigEmpty(t *testing.T) {
+	configTLSConfig := config.TLSConfig{
+		CAFile:             "",
+		CertFile:           "",
+		KeyFile:            "",
+		ServerName:         "",
+		InsecureSkipVerify: true}
+
+	expectedTLSConfig := &tls.Config{
+		InsecureSkipVerify: configTLSConfig.InsecureSkipVerify}
+	expectedTLSConfig.BuildNameToCertificate()
+
+	tlsConfig, err := NewTLSConfig(configTLSConfig)
+	if err != nil {
+		t.Fatalf("Can't create a new TLS Config from a configuration (%s).", err)
+	}
+
+	if !reflect.DeepEqual(tlsConfig, expectedTLSConfig) {
+		t.Fatalf("Unexpected TLS Config result: \n\n%+v\n expected\n\n%+v", tlsConfig, expectedTLSConfig)
+	}
+}
+
+var invalidTLSConfig = []struct {
```

Comment:
I used https://github.com/prometheus/prometheus/blame/master/config/config_test.go#L577 as inspiration, also is the pattern used in https://golang.org/src/fmt/fmt_test.go#L134. Of course I can change it if you want.